EN-30 Rename projects_topics to projects_allowed_input_topics

jinjagit opened the pull request:
## Links

- [citizenlab-ee PR](https://github.com/CitizenLabDotCo/citizenlab-ee/pull/158)
- [Epic](https://citizenlab.atlassian.net/browse/EN-1757)
- [Epic Deployment](https://en30.epic.citizenlab.co)

This PR prepares for the introduction of a new association of `topics` with `projects` by renaming the current `projects_topics` table to `projects_allowed_input_topics` and the Project model attribute `topics` to `allowed_input_topics`.

The plan is to have 2 different associations of `topics` with `projects`:

1. `projects_allowed_input_topics` = the pool of `topics` that can be used within a project to tag inputs (Ideas, Initiatives). Was called `projects_topics`. This renaming is the focus of this PR.
2. `projects_topics` = a new tagging of the project itself, drawing from the same platform-wide pool of topics that are available for (1), but being a different selection (probably fewer), and enabling projects to be filtered by topic (tag), for example. Note: this naming would be most likely to suggest (2) to a new dev, hence the renaming of the existing table + model attribute in this PR.

## How urgent is a code review?

By Tuesday (01/02/22) next week would be great, as this is the first step in the 3rd (final) iteration of this epic.

jinjagit committed:
migrate changes and rename many things

alexander-cit committed:
EN-30 Properly define associations in projects-topics join table

jinjagit committed:
Fixes to get specs green

Was 175 failures. Now no spec failures, when
run docker-compose run --rm web rspec

jinjagit committed:
Stub class in old migration

jinjagit committed:
stub class in historical migration + other small fixes

jinjagit committed:
fix clusterings_spec

jinjagit committed:
fix failing engine specs

jinjagit commented on back/db/migrate/20200527094026_add_ordering_to_projects_topics.rb:
```diff
@@ -1,10 +1,18 @@
+
+  def down
+   # projects_topics :ordering, :integer
+  end
```
I was trained to never change existing migrations. However, in this case the original migration referenced a class, and Rails would complain this class does not exist (which it doesn't at this point) when running `db:migrate`. Stubbing the class solves this problem.

Thanks to @alexander-cit for this solution.

sebastienhoorens replied:
I think for `down` the ordering and timestamps columns should be removed.

jinjagit commented on back/app/models/activity.rb:
```diff
@@ -13,9 +13,9 @@
-#  index_activities_on_acted_at               (acted_at)
-#  index_activities_on_item_type_and_item_id  (item_type,item_id)
-#  index_activities_on_user_id                (user_id)
+#  index_activities_on_acted_at  (acted_at)
+#  index_activities_on_item      (item_type,item_id)
+#  index_activities_on_user_id   (user_id)
```
This change is introduced when `db:migrate` is run.

Maybe we should leave this in the PR, so that this file change doesn't confuse others when running `db:migrate` locally?

sebastienhoorens replied:
Sure!

sebastienhoorens commented on back/app/controllers/web_api/v1/projects_allowed_input_topics_controller.rb:
```diff
@@ -0,0 +1,59 @@
+class WebApi::V1::ProjectsAllowedInputTopicsController < ApplicationController
+  before_action :set_projects_allowed_input_topic, only: %i[show reorder destroy]
+  skip_before_action :authenticate_user
```
I think this line could be improved. For example, when reordering or deleting, I think an authenticated user will be required?

jinjagit replied:
Yes, and actually also for creating.

jinjagit replied:
@sebastienhoorens Unfortunately, using this line:
```
skip_before_action :authenticate_user, only: %i[index, show]
```
Results in this spec failure:
```
rspec ./spec/acceptance/projects_allowed_input_topics_spec.rb:35 # ProjectsAllowedInputTopics GET web_api/v1/projects/:id/projects_allowed_input_topics List all projects_allowed_input_topics of a project
```
Which is exactly the unauthorized index view we hope is still permitted (but is actually getting a '401 - not authorized' response). I cannot figure out why this is the case.

jinjagit replied:
Now all controller actions (for this controller) require an active admin or project moderator (though policies: `ProjectsAllowedInputTopicPolicy`). Specs have been adjusted to reflect this.

sebastienhoorens commented on back/app/controllers/web_api/v1/projects_allowed_input_topics_controller.rb:
```diff
@@ -0,0 +1,59 @@
+
+  def index
+    @projects_allowed_input_topics = policy_scope(ProjectsAllowedInputTopic)
+    @projects_allowed_input_topics = @projects_allowed_input_topics.where(project_id: params[:project_id]) if params[:project_id].present?
```
Does this endpoint still makes sense when no project_id is provided? Otherwise, I think it's better to nest the route under projects. For example `GET /project/:project_id/allowed_input_topics`.

jinjagit replied:
I think you are right, it does not make sense to list all `allowed_input_topics` without also limiting the list to those related to a specific project. I have added a nested route (and removed `:index` from the `resources ... only: [ ... ]` list), to give:
```
resources :projects_allowed_input_topics, only: [:show, :create, :reorder, :destroy] do
   patch 'reorder', on: :member
   get '/projects/:project_id/projects_allowed_input_topics', on: :collection, to: 'projects_allowed_input_topics#index'
end
```
Is this what you advised?

jinjagit replied:
@sebastienhoorens Using nested routes, I was only able to reduced the _non_-nested routes to this:
```
resources :projects_allowed_input_topics, only: [:index, :show, :create, :destroy] do
    patch 'reorder', on: :member
end
```
`:index` is only required here to enable the request tested in the spec `get 'web_api/v1/projects_allowed_input_topics'` (which you advised to keep) in `back/spec/acceptance/projects_allowed_input_topics_spec.rb`, though I note I cannot see this endpoint in use in the FE (i.e. `front/app/services/projectTopics.ts` only uses this path for creation of a new item).  
  
`:create` is required here, probably because `front/app/services/projectTopics.ts` uses `${API_PATH}/projects_allowed_input_topics` for `addProjectTopic`

sebastienhoorens reviewed (approved):
Well done! 👏 I have a few remarks, but nothing truly blocking. We'll need to regenerate the external templates after releasing this, otherwise platform creation will be broken.

jinjagit committed:
Remove timestamps and ordering column when running down in migration

This ensure the effects of def up are achieved when def down is run.

jinjagit committed:
Enforce authorization for approriate controller actions

In ProjectsAllowedInputTopicsController, enforce authorization
for create, reorder and destroy, but not index or show.

jinjagit committed:
Edit routes to specifically expect project id for index action.

Nested route: get '/project/:project_id/allowed_input_topics',
on: :collection, to: 'projects_allowed_input_topics#index'

jinjagit committed:
Edit example namings in spec

jinjagit committed:
Comment out non-sensical spec.

This spec makes no sense, since there is no reason to list all
allowed_input_topics associated with all projects.
To delete after review by peer, confriming my resonaning.

jinjagit commented on back/spec/acceptance/projects_allowed_input_topics_spec.rb:
```diff
@@ -0,0 +1,105 @@
+  #     json_response = json_parse(response_body)
+  #     expect(json_response[:data].size).to eq 2
+  #   end
+  # end
```
@sebastienhoorens I am thinking of deleting this spec, since I can't think when it would make sense to list all `allowed_input_topics` for all projects.

I believe this PR also changes the index route to avoid this possibility in any case.

jinjagit committed:
match route with endpoint

jinjagit committed:
Improve route nesting

jinjagit committed:
All actions require active admin or moderator

Also improve routes
Related specs should now be passing, but need a moderator example
too for added permissions coverage.

jinjagit committed:
add one spec to test moderator permissions

jinjagit committed:
skip moderator spec when not ee

alexander-cit commented on back/app/policies/projects_allowed_input_topic_policy.rb:
```diff
@@ -12,6 +12,14 @@ def resolve
     end
   end
 
+  def index?
```
Why did you need to add it?

jinjagit replied:
To add permission for active admin or moderator.

alexander-cit replied:
I mean the ticket is just about renaming. Why did you change permissions? Was it a bug? An improvement?

jinjagit replied:
In discussion with Sebastien, we realised there is no reason a not-signed in user should be able to call any action in the `projects_allowed_inputs_topics` controller.

jinjagit replied:
An improvement. Sorry, I should have made that clear. It comes from discussion with @sebastienhoorens, following his review.

alexander-cit replied:
I see, it makes sense. Just one tiny improvement. As we add more and more methods with the same code here, it makes more and more sense to introduce a private method e.g.
```ruby
private

def active_admin_or_moderator?
  user&.active? && user&.active_admin_or_moderator?(record.project_id)
end
```
and call this method in public methods.

jinjagit replied:
This is done. Thanks!

jinjagit committed:
refactor to use private def

alexander-cit commented on back/db/migrate/20200527094026_add_ordering_to_projects_topics.rb:
```diff
@@ -1,10 +1,19 @@
     add_column :projects_topics, :ordering, :integer
 
-    ProjectsTopic.order(:created_at).each.with_index do |projects_topic, index|
+    StubProjectsTopic.order(:created_at).each.with_index do |projects_topic, index|
```
Good catch with `down` method. Indeed, rollback wouldn't work in this migration. To make it a bit simpler and more reliable, you can also leave `change` method (not splitting into up/down) and use reversible migration https://edgeguides.rubyonrails.org/active_record_migrations.html#using-reversible

jinjagit replied:
Done. Thanks, I didn't know about this method.

jinjagit committed:
Use 'reversible' in edited migration

alexander-cit commented on back/db/migrate/20200527094026_add_ordering_to_projects_topics.rb:
```diff
@@ -1,10 +1,21 @@
 
-    ProjectsTopic.order(:created_at).each.with_index do |projects_topic, index|
-      projects_topic.update_column :ordering, index
+    dir.down do
```
The main advantage of reversible is built-in rails migration methods will continue to work in both directions (migrate and rollback). So, it can look like that.

```ruby
class StubProjectsTopic < ActiveRecord::Base
  self.table_name = 'projects_topics'
end
  
def change
  add_timestamps :projects_topics
  add_column :projects_topics, :ordering, :integer

  reversible do |dir|
    dir.up do
      StubProjectsTopic.order(:created_at).each.with_index do |projects_topic, index|
        projects_topic.update_column :ordering, index
      end
    end
  end 
end
``` 

So, the original code is almost untouched.

jinjagit committed:
improve use of reverisble in migration

jinjagit committed:
remove useless only: value in route

alexander-cit commented on back/app/controllers/web_api/v1/projects_allowed_input_topics_controller.rb:
```diff
@@ -0,0 +1,59 @@
+class WebApi::V1::ProjectsAllowedInputTopicsController < ApplicationController
+  before_action :set_projects_allowed_input_topic, only: %i[show reorder destroy]
+  #skip_before_action :authenticate_user, only: %i[index, show]
```
It seems indeed this controller is not used by citizens, but still please try to test (first locally, then on staging) all projects/topics-related functionality after signing out.

jinjagit committed:
remove commented out code

alexander-cit commented on front/app/services/projectTopics.ts:
```diff
@@ -4,7 +4,7 @@ import { IRelationship } from 'typings';
 export interface IProjectTopicData {
   id: string;
-  type: 'projects_topic';
+  type: 'projects_allowed_input_topic';
```
It's great that you updated FE code without FE-devs help!

alexander-cit reviewed (approved):
It seems we already ran e2e tests, but maybe it's worth running them again in master after merging this PR. Definitely not mandatory, up to you :)
